node/engine/engine.go

```diff
@@ -189,6 +189,8 @@ func (e *Engine) run(ctx context.Context) {
 		case pr := <-e.PaymentRequestsFromAPI:
 			res, err = e.handlePaymentRequest(pr)
 		case chainEvent := <-e.fromChain:
+			err = e.store.SetLastBlockProcessed(chainEvent.BlockNum())
```

Comment:
Since we're setting `lastBlockProcessed` when the engine receives an event with that block number, it means `lastBlockProcessed` gets set when the engine **handles the first event in a block**, not when all events in a block are handled. 

For example lets say we have block which contains two events: 1)`Deposited 0x0, 0x0AAA... Block Num:55` 2)`Deposited 0x0, 0xBBBB...Block Num:55`

Let's say our engine receives the first event and sets `lastBlockProcessed=55`. Now let's say the engine crashes. Based on the `lastBlockProcessed` it has no way of knowing that we never handled the second event.

Reply:
See #1591

Reply:
> If we already have protection in the Channel class against processing the same event twice

I think we need an ADR for this - a well-thought-through pattern which involves some "contract" between a chain service and the `Channel` class, such that we don't get any bugs around stale or incorrect information.

Example: the `Channel` class promises to have an idempotent function for accepting events from the chain service, and the chain service promises to pass events in order. 

☝️ I'm not 100% sure about this example though. It seems like we might want to avoid (re)playing a `ChallengeRegistered` event if the challenge has long since expired or been cleared. 

Not a blocker for merging this PR, because I think in mose conceivable patterns we will want to store `lastBlockNumberSeen`.